src/main/java/emu/grasscutter/game/player/PlayerProgressManager.java

```diff
@@ -98,24 +81,34 @@ private void setOpenState(int openState, int value) {
     **********/
     private boolean areConditionsMet(OpenStateData openState) {
         // Check all conditions and test if at least one of them is violated.
-        for (var condition : openState.getCond()) {
+        for (OpenStateCond condition : openState.getCond()) {
             // For level conditions, check if the player has reached the necessary level.
             if (condition.getCondType() == OpenStateCondType.OPEN_STATE_COND_PLAYER_LEVEL) {
                 if (this.player.getLevel() < condition.getParam()) {
                     return false;
                 }
             }
-            else if (condition.getCondType() == OpenStateCondType.OPEN_STATE_COND_QUEST) {
-                // ToDo: Implement.
-            }
-            else if (condition.getCondType() == OpenStateCondType.OPEN_STATE_COND_PARENT_QUEST) {
-                // ToDo: Implement.
-            }
-            else if (condition.getCondType() == OpenStateCondType.OPEN_STATE_OFFERING_LEVEL) {
+
+            if (condition.getCondType() == OpenStateCondType.OPEN_STATE_OFFERING_LEVEL) {
                 // ToDo: Implement.
-            }
-            else if (condition.getCondType() == OpenStateCondType.OPEN_STATE_CITY_REPUTATION_LEVEL) {
+                return false;
+            } else if (condition.getCondType() == OpenStateCondType.OPEN_STATE_CITY_REPUTATION_LEVEL) {
                 // ToDo: Implement.
+                return false;
+            } else if (condition.getCondType() == OpenStateCondType.OPEN_STATE_COND_QUEST) {
+                // check sub quest id for quest finished met requirements
+                if (this.player.getQuestManager().getQuestById(condition.getParam()) == null
```

Comment:
When accessing a field that often, it's probably better to put it in a variable and then do the checks with that, instead of doing many  function calls each time.